osu.Game/Overlays/DirectOverlay.cs

```diff
@@ -225,28 +246,139 @@ protected override void PopIn()
             base.PopIn();
 
             // Queries are allowed to be run only on the first pop-in
-            if (getSetsRequest == null)
+            if (beatmapSetPager == null)
                 queueUpdateSearch();
         }
 
-        private SearchBeatmapSetsRequest getSetsRequest;
-
         private readonly Bindable<string> currentQuery = new Bindable<string>(string.Empty);
 
         private ScheduledDelegate queryChangedDebounce;
+        private ScheduledDelegate addPageDebounce;
         private PreviewTrackManager previewTrackManager;
+        private BeatmapSetPager beatmapSetPager;
+
+        private class BeatmapSetPager
+        {
+            public event PageFetchHandler PageFetch;
+
+            private readonly RulesetStore rulesets;
+
+            private readonly string query;
+            private readonly RulesetInfo ruleset;
+            private readonly BeatmapSearchCategory searchCategory;
+            private readonly DirectSortCriteria sortCriteria;
+            private readonly SortDirection sortDirection;
+
+            private SearchBeatmapSetsRequest getSetsRequest;
+
+            private int currentPage = 1;
+
+            public bool IsLastPageFetched { get; private set; } = false;
+            public bool IsFetching => getSetsRequest != null;
+
+            public BeatmapSetPager(RulesetStore rulesets, string query, RulesetInfo ruleset, BeatmapSearchCategory searchCategory = BeatmapSearchCategory.Any, DirectSortCriteria sortCriteria = DirectSortCriteria.Ranked, SortDirection sortDirection = SortDirection.Descending)
+            {
+                this.rulesets = rulesets;
+
+                this.query = query;
+                this.ruleset = ruleset;
+                this.searchCategory = searchCategory;
+                this.sortCriteria = sortCriteria;
+                this.sortDirection = sortDirection;
+            }
+
+            public SearchBeatmapSetsRequest FetchNextPage()
```

Comment:
Now that I'm reading this a bit more carefully I'm not sure about this API choice to return a request. It does a bit of a WTF thing if a request is in-flight (returns a `null`). Honestly I'd consider making the pager queue API requests itself and making this `void`. Or `bool`, at most, to get rid of some null-checks, and return `false` if there's an ongoing request.

Also this component is technically not thread-safe since you could get races between reads of `getSetsRequest` and the assignment, therefore getting >1 in-flight request where the last one gets assigned. Easiest (but maybe not best) way to fix is to add an exclusive lock on that critical section. I don't expect too much starvation, debounce logic would make hitting this race probably nigh impossible too, but let's do things properly.

Reply:
I share your concern about this function as well. I just wasn't sure when I was writing it what the proper way to lower the `API` provider into this component would be, and forgot to make a note about it. What's the best way to do this?

Reply:
Also, I suppose a lock would fix the race condition you mention. I'll try to think of an alternate solution, but this is probably the most C# I've ever written, so my odds of coming up with something better are low. :smile:

Reply:
I'd go with constructor injection. With the current instance-per-filter-set implementation you'd probably have to pass it to the ctor manually. You could also make the pager an always-present visual component with no content and use `[Resolved]` injection, but that's slightly uglier IMO.